docs/proposals/DNSPolicy.md

```diff
@@ -0,0 +1,409 @@
+
+
+## None Goals
+
+- Allow different policies for different listeners. Although this may be something we look to support in the future, currently policy attachment does not allow for this type of targeting. This means a DNSPolicy is applied for the whole gateway currently. 
+- Define how health checks should work, this will be part of a separate proposal
+
+
+## Terms
+
+- **managed listener**: This is a listener with a host backed by a DNS zone managed by the multi-cluster gateway controller
+- **hub cluster**: control plane cluster that managed 1 or more spokes
+- **spoke cluster**: a cluster managed by the hub control plane cluster. This is where gateway are instantiated
+
+## Proposal
+
+Provide a control plane DNSPolicy API that uses the idea of direct [policy attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/#direct-policy-attachment) from gateway API that allows a load balancing strategy and TTLs to be applied to the DNS records for any managed listeners being served by the data plane instances of this gateway. 
+The DNSPolicy also covers DNS based health checks but that is not covered in this document.
+
+Below is a draft API for what we anticipate the DNSPolicy to look like
+
+```yaml 
+
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSPolicy
+spec:
+  targetRef: # defaults to gateway gvk and currrent namespace
+    name: gateway-name
+  health:
+   ...
+  ttl: 
+    value: 60
+    gracePeriod: 600 #(seconds to wait beyond TTL before allowing deletion of gateway) 
+  loadBalancing:
+     #one or both of geo and weighted are required 
+    weighted:
+     default: 10  #required if weighted is chosen
+     custom:
+     - attribute: cloud  #optional with both geo and weighted
+       value: AWS
+       weight: 10
+     - attribute: cloud
+       value: GCP
+       weight: 20
+    geo:
+      default: IE # required if geo is chosen
+```  
+
+With the above API the two strategies are weighted or geo however you can use both to get a geo and weighted within a geo load balancing. By defining a geo section, you are indicating that you want to use a geo based strategy. A geo based strategy can be complimented with a weighted strategy. If you do not define a geo section, then you are choosing a purely weighted strategy.
+  
+### Load Balancing Strategies  
+
+Geo and Weighted load balancing are well understood strategies and this API effectively allow a complex requirement to be expressed relatively simply and executed by the gateway controller. Our default policy will execute a "Round Robin" weighted strategy which reflects the current behaviour.
+
+```yaml 
+
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSPolicy
+name: default-round-robin
+spec:
+  targetRef: # defaults to gateway gvk and currrent namespace
+    name: gateway-name
+  health:
+   ...
+  ttl: 
+    value: 60
+    gracePeriod: 240 #(seconds to wait beyond TTL before allowing deletion of gateway) 
```

Comment:
presumably this is a delay on how long we wait before removing an endpoint from the endpoint list rather than deleting a gateway?

Reply:
As I have understood this, the concept is that this is how long we will wait after the endpoint is removed from the DNSRecord, to remove the gateway from the cluster (i.e. allow time for the removed DNS Record endpoint to propagate).

Reply:
ok, i see. How are we going to block the deletion of a gateway based on something the dns policy is doing? I read the text below, but I'm still not understanding how it works since the dns policy won't directly affect a gateway other than maybe updating a status.

Reply:
yeah, that's what I'm trying to figure out right now.

Reply:
It seems arguably nothing to do with dns. You want to delay the deletion of the target by x amount, once dns is no longer enabled for that gateway. The only way i can you see you doing it is having the dns policy put a finalizer on the Gateway and I'm not sure if that is really what we want is it?

Reply:
Related to #138 (which I see assigned to @philbrookes )
If we agree on the grace period being solely related to keeping a gateway running for latent traffic, perhaps we remove it from DNSPolicy spec and figure out the right solution in #138?

It's more than gateway deletion that would need to be taken into account, so a finalizer may not work. 
If a gateway is 'moved' via placementdecision changes, that grace period should come into effect. 
Owning the finalizer of an individual gateway in a spoke cluster from the hub would have architectural considerations.

Reply:
It is related to DNS by the TTL but I see where you are coming from. So I guess the question is where does it belong?
As this only matters when the placement changes, and an existing cluster is removed from the placement decision, perhaps for now we look to add an annotation to the gateway itself? It may migrate into a policy at somepoint

Reply:
> It is related to DNS by the TTL

It's related to placement from what i understand, but wants to base a decision on some DNS config(TTL) . It doesn't directly have any influence on what the DNS reconciler will do.

The Gateway controller (our ocm one) can depend on a DNSPolicy and read from it if it wants, but "gracePeriod" for placement decisions probably still doesn't belong in it. The TTL itself can be read from the DNSRecord directly, since these still belong to a Gateway.

Reply:
so I think this sounds like we're discussing a gateway "DeschedulingPolicy" right? Where we want to say something like "if the DNS Record still exists for any gateway then onDeschedule of a particular gateway wait for the DNS Records TTL + some value in the policy"? Are we thinking then, that I will first implement this in code, and possibly later we'll pull it out to an actual policy, or shall I look into making the DeschedulingPolicy now?

Reply:
decision: use a code based default for now and allow it to be overriden on the gateway with an experimental annotation `kuadrant.io/experimental-grace-period`